### ISSUE TITLE
Can not use a PKCS#11 keystore when configuring the HTTPS listener

### ISSUE THREAD

**Issue by pedroigor** [quarkusio/quarkus#31172]
### Describe the bug

Looks like Quarkus is forcing either PEM files or a key store in order to start the HTTPS listener.

This seems to be a blocker if you want to use a PKCS#11 keystore where key material is loaded from the underlying PKCS#11 module (e.g.: some HSM device or soft-based PKCS#11 impl like `SoftHSM`). When using PKCS#11 there is no file to read from but rely on the `SunPKCS11` security provider configuration to integrate with the underlying cryptographic device.

After some investigation, I noticed that even though `ssl` is enabled [here](https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L908), right after the server options are set to `null` as per https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L694.

To workaround the issue, I had to create `HttpServerOptionsCustomizer` in order to create a PKCS#11 based `JksOptions` :

```
private static final String PKCS11_KEY_STORE_TYPE = "PKCS11";

    @ConfigProperty(name = "kc.https-key-store-type", defaultValue = "none")
    String httpsKeyStoreType;

    @ConfigProperty(name = "kc.https-key-store-password")
    String httpsKeyStorePassword;

    private KeyStoreHelper keyStoreHelper;

    @Override
    public void customizeHttpsServer(HttpServerOptions options) {
        if (PKCS11_KEY_STORE_TYPE.equalsIgnoreCase(httpsKeyStoreType)) {
            try {
                keyStoreHelper = new KeyStoreHelper(loadPkcs11KeyStore(), httpsKeyStorePassword, null);
                options.setKeyStoreOptions(createPkcs11KeyStoreOptions(keyStoreHelper));
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }
    }

    private KeyStore loadPkcs11KeyStore() {
        try {
            KeyStore ks = KeyStore.getInstance(httpsKeyStoreType);

            ks.load(null, Optional.ofNullable(httpsKeyStorePassword).map(String::toCharArray).orElse(null));

            return ks;
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private JksOptions createPkcs11KeyStoreOptions(KeyStoreHelper keyStoreHelper) {
        return new JksOptions() {
            @Override
            public JksOptions copy() {
                return this;
            }

            @Override
            public KeyManagerFactory getKeyManagerFactory(Vertx vertx) throws Exception {
                return keyStoreHelper.getKeyMgrFactory();
            }

            @Override
            public Function<String, X509KeyManager> keyManagerMapper(Vertx vertx) throws Exception {
                return keyStoreHelper::getKeyMgr;
            }
        };
    }
```

By doing that, Quarkus/Vert.x starts the HTTPS listener using my PKCS#11 keystore.

If the issue makes sense, I can try to come up with a PR.

### Expected behavior

Enable HTTPS using a PKCS#11 keystore.

### Actual behavior

HTTPS listener won't start if PEM files or key store is not provided.

### How to Reproduce?

1. Configure your environment as follows https://github.com/keycloak/keycloak/issues/17098#issuecomment-1430508939, so that you can use SoftHSM as a PKCS#11 complaint key store.
2. Change your Java security properties file (java_home/conf/security/java.security) to configure the `SunPKCS11` provider as follows:
```
# configuration for security providers 1-12 omitted
security.provider.13=SunPKCS11 ~/security.softhsm/tokens/softhsm-pkcs11.conf
```
3. Start an application using the following options:
```
quarkus.http.ssl.certificate.key-store-password=mypin
quarkus.http.ssl.certificate.key-store-file-type=PKCS11
```

### Output of `uname -a` or `ver`

Linux fedora 6.1.10-100.fc36.x86_64 #1 SMP PREEMPT_DYNAMIC Mon Feb  6 19:58:39 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.5" 2022-10-18 OpenJDK Runtime Environment Temurin-17.0.5+8 (build 17.0.5+8) OpenJDK 64-Bit Server VM Temurin-17.0.5+8 (build 17.0.5+8, mixed mode, sharing)

### Quarkus version or git rev

2.13.7

**Comment by geoand** [comment]
cc @cescoffier

**Comment by cescoffier** [comment]
It's a known issue. I need to redo the ssl config since almost 3 years now... never had enough time to do it.

**Comment by sberyozkin** [comment]
@pedroigor @cescoffier Makes sense to address it as the same issue would block Quarkus and FIPS integration for Vert.x HTTP as we discussed with Pablo.

@pedroigor I believe https://github.com/quarkusio/quarkus/blob/main/extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/VertxHttpRecorder.java#L874 is where the fix should be applied, you mentioned Vert.x will load natively from the default keystore (aka `null` as in your customizer) if no file is set, so if for example the essential store properties like password and type are both set then passing `null` to the keyStoreOptions should be OK

**Comment by cescoffier** [comment]
We would also need to track all the server and client to add the same feature (in Quarkus core and quarkiverse)

**Comment by sberyozkin** [comment]
@cescoffier Do you mean updating those libraries which support loading the keys from the keystores to load from the default/null ? Yes, sure, we worked with Pablo on getting smallrye-jwt/smallrye-jwt-build supporting loading from the null store, indeed, any other Quarkus extension dealing with loading the keys which would like to be FIPS compliant or just be able to load from the default store for whatever reasons, will have to be updated

**Comment by cescoffier** [comment]
@sberyozkin No, I mean that we have plenty of extension creating vert.x client and reproducing more or less the same pattern to configure SSL. If we add PKCS#11 in the server, we need to to it for all the servers (gRPC), and clients (gRPC client, rest client, the WebClient from vault....)

**Comment by cescoffier** [comment]
(which is why I want to move toward a "registry" based configuration with everything in a single place and extension referencing the "key set" they want)

**Comment by sberyozkin** [comment]
@cescoffier The config registry is a good idea, sure, it is not a pre-requisite for Vert.x HTTP being able to load from the default store. For example. Keycloak server may be doing MTLS FIPS with non-Quarkus HTTP clients.
Other extensions you've mentioned can be enhanced iteratively

**Comment by cescoffier** [comment]
Yes sure, but we tried that, and we ended up with many different configuration patterns, and heterogeneous support.

It might be time to do things right and stop patching everywhere. Not saying the registry should go in first (that would take several months of work), but at least adding the support in all the extensions using vertx servers and clients the SAME way.

**Comment by sberyozkin** [comment]
Checking if the keystore location is null does not change any configuration patterns. 

>but at least adding the support in all the extensions using vertx servers and clients the SAME way.

I don't think there is any contradiction in what I was saying with your point above, and specifically when I said it can be done iteratively. How else, a single PR enhancing everything in Quarkus at the same time is feasible but IMHO it is not necessary, perhaps you'd like to create an epic issue and updating Vert.x HTTP is the first tick in that epic, etc ?

**Comment by pedroigor** [comment]
@cescoffier @sberyozkin Thanks.

Vert.x should load the PKCS#11 store just fine. However, there is this issue https://github.com/eclipse-vertx/vert.x/pull/2626 which seems to provide a proper `KeyCertOptions` implementation specific for FIPS. Looks like it is blocked by how to properly test it (we are also running into the same problem).

In our case, we are also adding an option to make it easier to configure the `SunPKCS11` security provider similarly to https://quarkus.io/guides/security-customization#sun-pkcs11.

**Comment by pedroigor** [comment]
FYI, if you set a valid key store (without using that customizer) Vert.X won't use it to fetch keys but the PKCS11 store.

**Comment by sberyozkin** [comment]
@pedroigor Hey, using FIPS graded keys/algorithms is indirectly facilitated by configuring a keystore type/provider, as you've done with the customizer, or in the BC FIPS TLS test. I'm not sure having a FIPS specific branch is necessary, the requirement to read from PKCS11/default keystore might not be tied to FIPS but like you said to a more general requirement to read the keys from the hardware system etc and these keys may not be FIPS level keys

**Comment by Vincent-Le** [comment]
Hi all, does anyone have an update on this issue?  Or know if/when PKCS#11 will be supported.  We need FIPS compliant.

**Comment by cescoffier** [comment]
At the moment, the workaround is to use PKCS#12 (which, if I'm not mistaken, is also FIPS compliant). 

With the new TLS registry, an extension can register its own certificate (using their format of choice). It could also be a (more complicated) workaround.

**Comment by cescoffier** [comment]
Unfortunately, I was unable to test PKCS#11 with the TLS registry. Anyone interested in testing?
